sympy/core/function.py

```diff
@@ -613,6 +613,16 @@ def _eval_is_commutative(self):
     def _eval_is_complex(self):
         return fuzzy_and(a.is_complex for a in self.args)
 
+    def _eval_is_zero(self):
+        f = self.func(*self.args, evaluate=True)
+        if f.func != self.func:
+            return f.is_zero
```

Comment:
I'm not sure about this. I have generally approached this from the perspective that if the user doesn't want to evaluate then the assumptions system shouldn't do the evaluation. We should generally assume that the user had a good reason for not evaluating.

Reply:
The whole idea is really that I think all checks should be done as `.is_zero` rather than `== 0`. In the longer run, the functions can catch up to return better results.

Note that `Add(1, -1, evaluate=False).is_zero` is currently `True`.

Part of the motivation comes from #17454 and #17482 where the polynomial has a form of `x**2 + x*expr` but in a later stage it turned out that `expr` was 0. Hence, checking for `== 0` or `is S.Zero` is not really viable, but with a good `is_zero` things would be easier.

Also, consider `Sum(0, (x, M, N)). is_zero`, `sin(Symbol('x', zero=True))`, #17414, and the Zen of SymPy: "Unevaluated is better than evaluated.". If one would like to know if the expression has evaluated to `0` or `S.Zero` check against that. If one would like to know if the expression (for any reason) will evaluate or has evaluated to zero, check `is_zero`. As the assumptions can figure out things that the numerical evaluation cannot, it makes sense to primarily use that when possible in most code.

Reply:
Somehow, it is a question if the numerical comparisons should fallback to the assumptions or the assumptions should fallback to the numerical comparisons.

Reply:
Hence, I think that e.g. `sin` should have lines, somewhere, in `eval` that goes like:

```
if arg.is_zero:
    return S.Zero
```

as it right now doesn't benefit from the assumptions system in the evaluation. Maybe that is the way to solve/improve #17511 btw.

Reply:
So, basically, if you want to check that the unevaluated expression is 0, there are plenty of ways to do that without the assumption system.

Reply:
> The whole idea is really that I think all checks should be done as `.is_zero` rather than `== 0`

Given that in a parallel thread you are talking about mechanically changing the code to things that are faster I should point out the main difference here which is that is_zero can take an arbitrary amount of time. It should generally be quicker than doit (where doit might result in something more obviously zero) but it can be a lot slower than `== 0`.

```julia
In [17]: x = Symbol('x', positive=True)                                                                                           

In [18]: random_poly(x, 100, -10, 10) == 0                                                                                        
Out[18]: False

In [19]: random_poly(x, 100, -10, 10).is_zero  # Try this a few times because it can be fast or slow
```

Reply:
> I'm not sure about this.

I agree. I think `is_zero` is mainly there to give None/False values for expressions which are not identically zero. `Add(x, -x, evaluate=False).is_zero`  gives None and it should since `x - x` might be `nan` if `x` is not finite. `equals` should try to evaluate expressions, but assumptions should not. If the first check of `foo.is_zero` is `if foo is S.Zero: return True` then you can use `is_zero` instead of `is S.Zero` everywhere. But I would advise, with @oscarbenjamin , against trying to make the assumptions use evaluation.